src/Http/Http.Extensions/test/RequestDelegateFactoryTests.cs

```diff
@@ -2903,6 +2904,45 @@ public async Task RequestDelegateFactory_AsParameters_SupportsRequiredMember_Nul
     }
 #nullable enable
 
+    [ConditionalFact]
+    [RemoteExecutionSupported]
+    public void RequestDelegateFactory_WhenJsonIsReflectionEnabledByDefaultFalse()
+    {
+        var options = new RemoteInvokeOptions();
+        options.RuntimeConfigurationOptions.Add("System.Text.Json.JsonSerializer.IsReflectionEnabledByDefault", false.ToString());
+
+        using var remoteHandle = RemoteExecutor.Invoke(static () =>
+        {
+            // Arrange
+            var @delegate = (string task) => new Todo();
+
+            // IsReflectionEnabledByDefault defaults to `false` when `PublishTrimmed=true`. For these scenarios, we
+            // expect users to configure JSON source generation as instructed in the `NotSupportedException` message.
+            var exception = Assert.Throws<NotSupportedException>(() => RequestDelegateFactory.Create(@delegate));
+            Assert.Equal(
+                "JsonTypeInfo metadata for type 'Microsoft.AspNetCore.Routing.Internal.RequestDelegateFactoryTests+Todo' was not provided by TypeInfoResolver of type '[]'. If using source generation, ensure that all root types passed to the serializer have been annotated with 'JsonSerializableAttribute', along with any types that might be serialized polymorphically.",
```

Comment:
(nit) do we want to hard-code this whole exception message in our tests? If they update the message, our tests would break. Would it be better to assert that it contains a few of type names we know it should?

`Microsoft.AspNetCore.Routing.Internal.RequestDelegateFactoryTests+Todo`
`JsonSerializableAttribute`

?